File: src/data/personnel.ts
import type { Worker } from '../types'

const S = (a: Worker['status'][]) => a

export const workers: Worker[] = [
  { id: 'w1', name: '山田 太郎', org: '施工管理部 第一課', crew: '第一班', role: '工事長', licenses: ['職長・安全衛生責任者', '光ファイバ融着', '普通自動車免許'], assignedTo: '熊本中央局 光設備更改工事', schedule: S(['稼働', '稼働', '稼働', '稼働', '稼働', '休暇', '休暇']), status: '稼働', continuousDays: 5, vacation: '—', note: '現場責任者' },
  { id: 'w2', name: '田中 一郎', org: '施工管理部 第一課', crew: '第一班', role: '現場責任者', licenses: ['高所作業車', '玉掛け', '小型移動式クレーン', '普通自動車免許'], assignedTo: '菊陽町 通信管路敷設工事', schedule: S(['稼働', '稼働', '稼働', '稼働', '稼働', '稼働', '休暇']), status: '稼働', continuousDays: 6, vacation: '—', note: '連続勤務注意' },
  { id: 'w3', name: '高橋 誠', org: '施工管理部 第二課', crew: '第二班', role: '技術者', licenses: ['光ファイバ融着', '低圧電気取扱', 'フルハーネス特別教育'], assignedTo: '熊本中央局 光設備更改工事', schedule: S(['稼働', '稼働', '稼働', '稼働', '稼働', '待機', '休暇']), status: '稼働', continuousDays: 5, vacation: '—', note: '融着・測定担当' },
  { id: 'w4', name: '鈴木 健', org: '施工管理部 第三課', crew: '第三班', role: '技術者', licenses: ['電気工事士', 'フルハーネス特別教育', '普通自動車免許'], assignedTo: '合志市 基地局設備更新工事', schedule: S(['稼働', '稼働', '移動中', '稼働', '稼働', '休暇', '休暇']), status: '移動中', continuousDays: 3, vacation: '—', note: '2現場を兼務' },
  { id: 'w5', name: '佐藤 花子', org: '施工管理部 第二課', crew: '第二班', role: '現場責任者', licenses: ['職長・安全衛生責任者', '光ファイバ融着', '普通自動車免許'], assignedTo: '八代エリア FTTH増設工事', schedule: S(['稼働', '稼働', '稼働', '稼働', '稼働', '休暇', '休暇']), status: '稼働', continuousDays: 5, vacation: '—', note: '' },
  { id: 'w6', name: '伊藤 直樹', org: '施工管理部 第一課', crew: '第一班', role: '技術者', licenses: ['高所作業車', '玉掛け', '低圧電気取扱'], assignedTo: '熊本市東区 光ケーブル切替工事', schedule: S(['稼働', '稼働', '稼働', '稼働', '稼働', '待機', '休暇']), status: '稼働', continuousDays: 4, vacation: '—', note: '切替作業担当' },
  { id: 'w7', name: '渡辺 修', org: '施工管理部 第三課', crew: '第三班', role: '現場責任者', licenses: ['職長・安全衛生責任者', '酸素欠乏危険作業', '普通自動車免許'], assignedTo: '天草地区 通信設備復旧工事', schedule: S(['稼働', '稼働', '稼働', '稼働', '稼働', '稼働', '待機']), status: '稼働', continuousDays: 7, vacation: '要調整', note: '連続7日 過剰勤務警告' },
  { id: 'w8', name: '中村 亮', org: '施工管理部 第二課', crew: '第二班', role: '技術者', licenses: ['高所作業車', 'フルハーネス特別教育', '普通自動車免許'], assignedTo: '待機', schedule: S(['待機', '待機', '稼働', '稼働', '稼働', '休暇', '休暇']), status: '待機', continuousDays: 0, vacation: '—', note: '配置可能' },
  { id: 'w9', name: '小林 大輔', org: '協力会社 九州テクノ', crew: '応援', role: '作業員', licenses: ['光ファイバ融着', '玉掛け'], assignedTo: '待機', schedule: S(['待機', '稼働', '稼働', '稼働', '稼働', '休暇', '休暇']), status: '待機', continuousDays: 0, vacation: '—', note: '融着応援可' },
  { id: 'w10', name: '加藤 隆', org: '協力会社 肥後設備', crew: '応援', role: '作業員', licenses: ['高所作業車', '交通誘導'], assignedTo: '菊陽町 通信管路敷設工事', schedule: S(['稼働', '稼働', '稼働', '待機', '稼働', '休暇', '休暇']), status: '稼働', continuousDays: 3, vacation: '—', note: '' },
  { id: 'w11', name: '吉田 昇', org: '施工管理部 第一課', crew: '第一班', role: '作業員', licenses: ['玉掛け', '普通自動車免許'], assignedTo: '熊本中央局 光設備更改工事', schedule: S(['稼働', '稼働', '稼働', '稼働', '稼働', '休暇', '休暇']), status: '稼働', continuousDays: 5, vacation: '—', note: '' },
  { id: 'w12', name: '松本 康', org: '施工管理部 第三課', crew: '第三班', role: '作業員', licenses: ['酸素欠乏危険作業', '普通自動車免許'], assignedTo: '休暇', schedule: S(['休暇', '休暇', '稼働', '稼働', '稼働', '稼働', '休暇']), status: '休暇', continuousDays: 0, vacation: '7/21-7/22', note: '有給休暇' },
]

export const scheduleDates = ['7/21(月)', '7/22(火)', '7/23(水)', '7/24(木)', '7/25(金)', '7/26(土)', '7/27(日)']

File: backend/app/api/personnel.py
import json

from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.core.audit import write_audit
from app.core.db import get_db
from app.core.deps import ensure_project_access, get_current_user, require_roles
from app.models import (
    Company,
    Project,
    Qualification,
    Task,
    Team,
    User,
    Worker,
    WorkerAssignment,
    WorkerQualification,
)
from app.schemas import (
    AssignmentCheckRow,
    WorkerAssignmentCreate,
    WorkerAssignmentOut,
    WorkerDetailOut,
    WorkerOut,
    WorkerQualificationOut,
)

router = APIRouter()


def _schedule(w: Worker) -> list[str]:
    if not w.schedule:
        return []
    try:
        return json.loads(w.schedule)
    except (json.JSONDecodeError, TypeError):
        return []


def _licenses(db: Session, worker_id: int) -> list[str]:
    stmt = (
        select(Qualification.name)
        .join(WorkerQualification, WorkerQualification.qualification_id == Qualification.id)
        .where(WorkerQualification.worker_id == worker_id)
    )
    return list(db.execute(stmt).scalars().all())


def _assigned_to(db: Session, w: Worker) -> str | None:
    if w.status in ("待機", "休暇"):
        return w.status
    a = db.execute(
        select(WorkerAssignment).where(WorkerAssignment.worker_id == w.id).order_by(WorkerAssignment.id.desc())
    ).scalars().first()
    if a:
        proj = db.get(Project, a.project_id)
        return proj.name if proj else w.status
    return w.status


def _team_name(db: Session, w: Worker) -> str | None:
    if not w.team_id:
        return None
    t = db.get(Team, w.team_id)
    return t.name if t else None


def to_out(db: Session, w: Worker) -> WorkerOut:
    return WorkerOut(
        id=w.id, name=w.name, org=w.org, crew=_team_name(db, w), role=w.role,
        licenses=_licenses(db, w.id), assignedTo=_assigned_to(db, w), schedule=_schedule(w),
        status=w.status, continuousDays=w.continuous_days, vacation=w.vacation, note=w.note,
    )


@router.get("", response_model=list[WorkerOut])
def list_workers(
    project_id: int | None = Query(None, description="案件に配置された要員だけに絞り込む"),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> list[WorkerOut]:
    """要員一覧。案件配下の画面からは project_id を渡し、その案件の配置だけを返す。"""
    stmt = select(Worker).where(Worker.deleted_at.is_(None)).order_by(Worker.id)
    if project_id is not None:
        # 案件スコープと権限はここで強制する（画面の絞り込みに任せない）
        ensure_project_access(db, user, project_id)
        stmt = stmt.where(Worker.id.in_(
            select(WorkerAssignment.worker_id).where(WorkerAssignment.project_id == project_id)
        ))
    return [to_out(db, w) for w in db.execute(stmt).scalars().all()]


@router.get("/{worker_id}", response_model=WorkerDetailOut)
def get_worker(worker_id: int, db: Session = Depends(get_db), user: User = Depends(get_current_user)) -> WorkerDetailOut:
    w = db.get(Worker, worker_id)
    if not w or w.deleted_at is not None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "要員が見つかりません")
    base = to_out(db, w)
    company = db.get(Company, w.company_id) if w.company_id else None
    quals = db.execute(
        select(WorkerQualification, Qualification)
        .join(Qualification, Qualification.id == WorkerQualification.qualification_id)
        .where(WorkerQualification.worker_id == worker_id)
    ).all()
    qual_out = [
        WorkerQualificationOut(name=q.name, acquired_at=wq.acquired_at, expires_at=wq.expires_at, certificate_no=wq.certificate_no)
        for wq, q in quals
    ]
    assigns = db.execute(
        select(WorkerAssignment).where(WorkerAssignment.worker_id == worker_id).order_by(WorkerAssignment.id)
    ).scalars().all()
    assign_out = []
    for a in assigns:
        proj = db.get(Project, a.project_id)
        task = db.get(Task, a.task_id) if a.task_id else None
        assign_out.append(WorkerAssignmentOut(
            project_id=a.project_id, project=proj.name if proj else None,
            task_id=a.task_id, task=task.name if task else None,
            assigned_from=a.assigned_from, assigned_to=a.assigned_to, role=a.role, status=a.status,
        ))
    return WorkerDetailOut(
        **base.model_dump(), company=company.name if company else None,
        team=_team_name(db, w), qualifications=qual_out, assignments=assign_out,
    )


@router.post("/{worker_id}/assign", response_model=WorkerDetailOut)
def assign_worker(
    worker_id: int,
    body: WorkerAssignmentCreate,
    db: Session = Depends(get_db),
    user: User = Depends(require_roles("PROJECT_MANAGER")),
) -> WorkerDetailOut:
    """要員を案件（および工程）へ配置。配置すると稼働扱いにする。"""
    w = db.get(Worker, worker_id)
    if not w or w.deleted_at is not None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "要員が見つかりません")
    ensure_project_access(db, user, body.project_id)
    a = WorkerAssignment(
        worker_id=worker_id, project_id=body.project_id, task_id=body.task_id,
        assigned_from=body.assigned_from, assigned_to=body.assigned_to, role=body.role or w.role, status="稼働",
    )
    db.add(a)
    w.status = "稼働"
    db.flush()
    write_audit(db, user, "CREATE", "worker_assignment", a.id, project_id=body.project_id,
                after={"worker_id": worker_id, "task_id": body.task_id})
    db.commit()
    return get_worker(worker_id, db, user)


@router.delete("/{worker_id}/assign/{assignment_id}", response_model=WorkerDetailOut)
def unassign_worker(
    worker_id: int,
    assignment_id: int,
    db: Session = Depends(get_db),
    user: User = Depends(require_roles("PROJECT_MANAGER")),
) -> WorkerDetailOut:
    """配置解除。他の配置が無ければ待機に戻す。"""
    a = db.get(WorkerAssignment, assignment_id)
    if not a or a.worker_id != worker_id:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "配置が見つかりません")
    ensure_project_access(db, user, a.project_id)
    pid = a.project_id
    db.delete(a)
    db.flush()
    remaining = db.execute(select(WorkerAssignment).where(WorkerAssignment.worker_id == worker_id)).scalars().first()
    if not remaining:
        w = db.get(Worker, worker_id)
        if w:
            w.status = "待機"
    write_audit(db, user, "DELETE", "worker_assignment", assignment_id, project_id=pid, before={"worker_id": worker_id})
    db.commit()
    return get_worker(worker_id, db, user)


@router.get("/assignment-check/{project_id}", response_model=list[AssignmentCheckRow])
def assignment_check(
    project_id: int, db: Session = Depends(get_db), user: User = Depends(get_current_user)
) -> list[AssignmentCheckRow]:
    """工程ごとに 予定人数 と 配置人数 の整合を確認。"""
    ensure_project_access(db, user, project_id)
    tasks = db.execute(
        select(Task).where(Task.project_id == project_id, Task.deleted_at.is_(None), Task.wbs_code.like("%.%"))
    ).scalars().all()
    rows = []
    for t in tasks:
        cnt = db.execute(
            select(func.count(WorkerAssignment.id)).where(WorkerAssignment.project_id == project_id, WorkerAssignment.task_id == t.id)
        ).scalar() or 0
        rows.append(AssignmentCheckRow(task_id=t.id, task_name=t.name, planned_workers=t.planned_workers,
                                       assigned_count=cnt, ok=cnt <= t.planned_workers))
    return rows
